js/src/forum/components/EditUserModal.js

```diff
@@ -46,16 +46,23 @@ export default class EditUserModal extends Modal {
   fields() {
     const items = new ItemList();
 
-    items.add(
-      'username',
-      <div className="Form-group">
-        <label>{app.translator.trans('core.forum.edit_user.username_heading')}</label>
-        <input className="FormControl" placeholder={extractText(app.translator.trans('core.forum.edit_user.username_label'))} bidi={this.username} />
-      </div>,
-      40
-    );
+    if (app.session.user.canEditUsername()) {
+      items.add(
+        'username',
+        <div className="Form-group">
+          <label>{app.translator.trans('core.forum.edit_user.username_heading')}</label>
+          <input
+            className="FormControl"
+            placeholder={extractText(app.translator.trans('core.forum.edit_user.username_label'))}
+            bidi={this.username}
+          />
+        </div>,
+        40
+      );
+    }
 
-    if (app.session.user !== this.props.user) {
+    // The first part prevents silent account hijacking
```

Comment:
I don't understand what the account hijacking would be here.

You mean you'd want to prevent an admin from changing their password without sending themselves a reset email ? If so this should be controlled at the Policy level and blocked in the API as well. Just adding this check in the frontend does not prevent the bad actor from directly calling the API endpoint.

But I'd argue that even implementing this in the backend makes little sense. If you hacked into an admin account and want to discretely change the password, you could just as well change the password of another admin account, enter that other account, then change the password of your original hacked account.

Or you could just change their email first, which triggers no notification on the current email, then change password once a new email has been verified.

A better measure against account hijacking could be to send emails on changes. Like "Your email has been changed to <x>" to the old email and "Your password has been changed by an administrator" when an admin changes emails. But that would be another feature that could even be delegated to an extension.

Reply:
This was pre-existing based on the way I read the code I understood this as a way to prevent someone who hijacked an admin/moderator accounts from being able to change that admins/moderators account without the moderator from knowing. I completely agree that doing it in the fronted only isn't that great. We could remove that part all together.

Reply:
Clark brings up a good point that there's really no way to prevent hijacking if someone gets into an admin account. I think the strategy of notifying on email change is a good idea, but yeah I would side with removing that check and making sure the proper events are dispatched so that extensions can tune in.

Reply:
For now I've removed the comment, if I remove the check it allows admin and moderators to change their own password and email via the edit interface instead of settings. I'm not sure if that's something we want. (I do understand that via API it's still technically possible)

Reply:
I feel like it's even more confusing without a comment, as the check isn't really self-explanatory.

I really think this test should be removed from the frontend.

As for adding it to the backend instead, I don't think it makes sense either, as described in my first post.

Reply:
Removed

Reply:
Actually re-reading the code I think I know why that check was there in the first place. I don't think it had anything to do with security.

By forcing a user to use the email change feature, this makes sure they set a valid email.

If an administrator manually changes their own email via the edit user modal and put a wrong email by mistake, they might lock themselves out of their account.

I still prefer we get rid of the test (as has been done). If you have that power over other users, I think you can responsibly use it on your own account.

A text that recommends using the email change feature over manually setting an email could be a good idea maybe?

Reply:
@clarkwinkelmann Good idea, I'll see if I can get that working tonight maybe after work.

Reply:
If I may suggest, we could ship this PR without any special warning or feature and open an issue to keep track of that other possibility.

Actually I'd be very much in favor of changing the UI so that when you edit the email you get two options: either request an email change on the behalf of the user (default) or force-change the email (not default, with warning text)

Reply:
I agree, I'll create a new issue for the proposal and we can ship this PR without it